test/functional/km/control.cpp

```diff
@@ -0,0 +1,489 @@
+        //
+        if (!TestSetup()) {
+            TraceError(
+                "[ lib] ERROR, %s.",
+                "TestSetup failed");
+            TestDrvClient = nullptr;
+            Status = STATUS_UNSUCCESSFUL;
+            break;
+        }
+    }
+    while (false);
+
+    if (!NT_SUCCESS(Status) && Status != STATUS_TOO_MANY_SESSIONS) {
+        InterlockedExchange8(&TestDrvClientActive, FALSE);
+    }
+
+    WdfRequestComplete(Request, Status);
+}
+
+PAGEDX
+_Use_decl_annotations_
+VOID
+TestDrvCtlEvtFileClose(
+    _In_ WDFFILEOBJECT /* FileObject */
+    )
+{
+    PAGED_CODE();
+}
+
+PAGEDX
+_Use_decl_annotations_
+VOID
+TestDrvCtlEvtFileCleanup(
+    _In_ WDFFILEOBJECT FileObject
+    )
+{
+    PAGED_CODE();
+
+    TEST_CLIENT* Client = TestDrvCtlGetFileContext(FileObject);
+    if (Client != nullptr) {
+
+        TraceInfo(
+            "[test] Client %p cleaning up",
+            Client);
+
+        //
+        // TestSetup requires TestDrvClient to be initialized.
+        //
+        TestCleanup();
+
+        TestDrvClient = nullptr;
+
+        InterlockedExchange8(&TestDrvClientActive, FALSE);
+    }
+}
+
+VOID
+TestDrvCtlEvtIoCanceled(
+    _In_ WDFQUEUE /* Queue */,
+    _In_ WDFREQUEST Request
+    )
+{
+    NTSTATUS Status;
+
+    WDFFILEOBJECT FileObject = WdfRequestGetFileObject(Request);
+    if (FileObject == nullptr) {
+        Status = STATUS_DEVICE_NOT_READY;
+        goto error;
+    }
+
+    TEST_CLIENT* Client = TestDrvCtlGetFileContext(FileObject);
+    if (Client == nullptr) {
+        Status = STATUS_DEVICE_NOT_READY;
+        goto error;
+    }
+
+    TraceWarn(
+        "[test] Client %p canceled request %p",
+        Client,
+        Request);
+
+    Status = STATUS_CANCELLED;
+
+error:
+
+    WdfRequestComplete(Request, Status);
+}
+
+size_t IOCTL_BUFFER_SIZES[] =
+{
+    0,
+    0,
+};
+
+static_assert(
+    MAX_IOCTL_FUNC_CODE + 1 == (sizeof(IOCTL_BUFFER_SIZES)/sizeof(size_t)),
+    "IOCTL_BUFFER_SIZES must be kept in sync with the IOCTLs");
+
+typedef union {
+} IOCTL_PARAMS;
+
+#define TestDrvCtlRun(X) \
+    Client->TestFailure = false; \
+    X; \
+    Status = Client->TestFailure ? STATUS_FAIL_FAST_EXCEPTION : STATUS_SUCCESS;
+
+VOID
+TestDrvCtlEvtIoDeviceControl(
+    _In_ WDFQUEUE /* Queue */,
+    _In_ WDFREQUEST Request,
+    _In_ size_t /* OutputBufferLength */,
+    _In_ size_t InputBufferLength,
+    _In_ ULONG IoControlCode
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    WDFFILEOBJECT FileObject = nullptr;
+    TEST_CLIENT* Client = nullptr;
+
+    if (KeGetCurrentIrql() > PASSIVE_LEVEL) {
+        Status = STATUS_NOT_SUPPORTED;
+        TraceError(
+            "[ lib] ERROR, %s.",
+            "IOCTL not supported greater than PASSIVE_LEVEL");
+        goto Error;
+    }
+
+    FileObject = WdfRequestGetFileObject(Request);
+    if (FileObject == nullptr) {
+        Status = STATUS_DEVICE_NOT_READY;
+        TraceError(
+            "[ lib] ERROR, %s.",
```

Comment:
i don't think we need "ERROR" in each error message - we can add that to the `TraceError` macro if we want it later.